librbd: group snapshots

VictorDenisov opened the pull request.

VictorDenisov commented:
@dillaman This PR lives on top of yet unmerged snapshot namespaces PR. It doesn't have tests yet and snapshot remove operation. Could you please take a look and say if you like it in principle. Specifically this part: https://github.com/ceph/ceph/pull/11544/files#diff-d2dae46d02584254961fdcc4318289caR411
Once snapshot namespaces PR is merged I'll rebase this PR on top of master.

VictorDenisov commented:
Now the PR has create, remove, list operations. It's been rebased on top of master. It still lacks tests. I'm working on the tests.

dillaman commented on src/cls/rbd/cls_rbd.cc:
```diff
@@ -4797,6 +4803,246 @@ int image_get_group(cls_method_context_t hctx,
+ * Output:
+ * @return 0 on success, negative error code on failure
+ */
+int group_snap_save(cls_method_context_t hctx,
```
Minor: probably better named as ```group_snap_add```

dillaman commented on src/cls/rbd/cls_rbd.cc:
```diff
@@ -4797,6 +4803,246 @@ int image_get_group(cls_method_context_t hctx,
+    return -EINVAL;
+  }
+
+  std::string key = gs.snap_key();
```
Probably shouldn't allow all fields to be updated (e.g. the uuid shouldn't be updatable).

dillaman commented on src/cls/rbd/cls_rbd.cc:
```diff
@@ -4797,6 +4803,246 @@ int image_get_group(cls_method_context_t hctx,
+ * Retrieve snapshot record.
+ *
+ * Input:
+ * @param name Snapshot name
```
Probably should be the uuid since you should be able to rename group snaps.

dillaman commented on src/cls/rbd/cls_rbd.cc:
```diff
@@ -4797,6 +4803,246 @@ int image_get_group(cls_method_context_t hctx,
+  }
+
+  cls::rbd::GroupSnapshot gs;
+
```
TODO?

dillaman commented on src/cls/rbd/cls_rbd.cc:
```diff
@@ -4797,6 +4803,246 @@ int image_get_group(cls_method_context_t hctx,
+ * Remove snapshot record.
+ *
+ * Input:
+ * @param id Snapshot id
```
Use the uuid

dillaman commented on src/cls/rbd/cls_rbd.cc:
```diff
@@ -4797,6 +4803,246 @@ int image_get_group(cls_method_context_t hctx,
+
+  do {
+    keys_read = cls_cxx_map_get_vals(hctx, last_read,
+				     RBD_SNAP_KEY_PREFIX, // TODO  replace with prefix from start after
```
TODO

dillaman commented on src/cls/rbd/cls_rbd.cc:
```diff
@@ -4797,6 +4803,246 @@ int image_get_group(cls_method_context_t hctx,
+  do {
+    keys_read = cls_cxx_map_get_vals(hctx, last_read,
+				     RBD_SNAP_KEY_PREFIX, // TODO  replace with prefix from start after
+				     max_read, &vals);
```
Minor: max_read should take into account the number of keys already in ```res``` and ```max_return```

dillaman commented on src/cls/rbd/cls_rbd.cc:
```diff
@@ -4797,6 +4803,246 @@ int image_get_group(cls_method_context_t hctx,
+  int max_read = RBD_MAX_KEYS_READ;
+  std::map<string, bufferlist> vals;
+  string last_read = start_after.snap_key();
+  std::vector<cls::rbd::GroupSnapshot> res;
```
Minor: ```res``` -> ```group_snapshots```

dillaman commented on src/cls/rbd/cls_rbd_client.cc:
```diff
@@ -1770,5 +1770,96 @@ namespace librbd {
+    int group_snap_list(librados::IoCtx *ioctx, const std::string &oid,
+			const cls::rbd::GroupSnapshot &start,
+			uint64_t max_return,
+			std::vector<cls::rbd::GroupSnapshot>& snapshots) {
```
Minor: out params should be pointers not references

dillaman commented on src/cls/rbd/cls_rbd_types.h:
```diff
@@ -310,6 +310,55 @@ struct SnapshotNamespaceOnDisk {
+
+inline void encode(const GroupSnapshotState &state, bufferlist& bl,
+		   uint64_t features=0)
+{
```
Minor: you can move this to the previous line

dillaman commented on src/cls/rbd/cls_rbd_types.h:
```diff
@@ -310,6 +310,55 @@ struct SnapshotNamespaceOnDisk {
+  std::string name;
+  GroupSnapshotState state;
+
+  vector<ImageSnapshotRef> snaps;
```
At first glance, I am not sure we need two-way linkage between the group snapshot and the associate image snapshot. The image's snapshot already has the link back to the group snapshot and the group has links to all of the associated images. Therefore, I can iterate through all the images in the group to get the associated snapshots. I think we only need this if there is some high performance need to bypass the extra lookup.

dillaman commented on src/cls/rbd/cls_rbd_types.h:
```diff
@@ -310,6 +310,55 @@ struct SnapshotNamespaceOnDisk {
+struct GroupSnapshot {
+  static const uint64_t invalid_id = -1;
+  uint64_t id = invalid_id;
+  std::string uuid;
```
Minor: would be good to keep the naming consistent. ```GroupSnapshotNamespace``` refers to ```group_id``` which should be this ```uuid```, so I'd recommend renaming to ```id``` (since the sequence id isn't needed)

VictorDenisov replied:
little clarification request: does it mean want to have something like this:
struct GroupSnapshot {
   std::string id;

dillaman replied:
Yes, please.

dillaman commented on src/cls/rbd/cls_rbd.cc:
```diff
@@ -4797,6 +4803,246 @@ int image_get_group(cls_method_context_t hctx,
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
```
Need some basic validation of the provided group snapshot record. UUID empty, name empty, duplicate name, etc. Same comment applies to the update method.

dillaman commented on src/include/rbd/librbd.h:
```diff
@@ -159,6 +159,11 @@ typedef struct {
 } rbd_group_spec_t;
 
 typedef enum {
+  GROUP_SNAP_STATE_PENDING,
```
Don't forget to add the C and Python API methods to match the new C++ API methods

dillaman commented on src/include/rbd/librbd.hpp:
```diff
@@ -173,6 +180,13 @@ class CEPH_RBD_API RBD
+			const char *snap_name);
+  int group_snap_remove(IoCtx& io_ctx, const char *group_name,
+			const char *snap_name);
+  int group_snap_list(IoCtx& group_ioctx, const char *group_name,
```
Minor: should have support for rename as well.

VictorDenisov replied:
Rename will be addressed in a separate PR.

VictorDenisov replied:
https://github.com/ceph/ceph/pull/12431

dillaman replied:
Yup -- that was a very old comment.

dillaman commented on src/librbd/Group.cc:
```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
   return 0;
 }
+
+int group_snap_name_check_duplicate(librados::IoCtx& group_ioctx,
```
Should be enforced in cls_rbd not here

dillaman commented on src/librbd/Group.cc:
```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+  std::vector<librbd::IoCtx*> io_ctxs;
+  std::vector<librbd::ImageCtx*> ictxs;
+  std::vector<C_SaferCond*> on_finishes;
+  for (auto i: images) {
```
Minor: ```i``` -> ```image```

dillaman commented on src/librbd/Group.cc:
```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+  std::vector<librbd::ImageCtx*> ictxs;
+  std::vector<C_SaferCond*> on_finishes;
+  for (auto i: images) {
+    librbd::IoCtx* image_io_ctx = new librbd::IoCtx;
```
Minor: include ```()``` at end here and below for C_SaferCond

dillaman commented on src/librbd/Group.cc:
```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+      ldout(cct, 1) << "Failed to create io context for image" << dendl;
+    }
+
+    librbd::ImageCtx* image_ctx = new ImageCtx(i.name.c_str(), "", "",
```
"snap_name" param should be nullptr -- not ""

dillaman commented on src/librbd/Group.cc:
```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+    goto cleanup;
+  }
+  for (auto i: ictxs) {
+    i->exclusive_lock->block_requests(-EBUSY);
```
Need to hold ```owner_lock``` when touching ```ImageCtx::exclusive_lock``` -- and should verify that it isn't a nullptr before dereferencing. Same comment applies below.

dillaman commented on src/librbd/Group.cc:
```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+  if (ret_code != 0) {
+    goto cleanup;
+  }
+  for (auto i: ictxs) {
```
Minor: ```i``` -> ```image_ctx``` or ```ictx``` (use variable names w/ meaning). Comment applies throughout.

dillaman commented on src/librbd/Group.cc:
```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+  if (r < 0) {
+    return r;
+  }
+  int n = images.size();
```
Minor: ```n``` -> ```image_count``` or something similar

dillaman commented on src/librbd/Group.cc:
```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+  }
+
+  gs.id = snap_seq;
+  gs.uuid = "";
```
Generate a UUID

dillaman commented on src/librbd/Group.cc:
```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+
+  image_snaps = vector<cls::rbd::ImageSnapshotRef>(n, cls::rbd::ImageSnapshotRef());
+
+  ind_snap_name = calc_ind_image_snap_name(group_ioctx.get_id(),
```
Can just re-use the group snapshot uuid -- since it's guaranteed to be unique within the group namespace

dillaman commented on src/librbd/Group.cc:
```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+
+    C_SaferCond* on_finish = new C_SaferCond;
+
+    ictx->operations->snap_create(ind_snap_name.c_str(), ne, on_finish);
```
Just though about this: the ```add_snap``` method in cls_rbd should validate that the image is a member of the associated group before permitting the creation of a ```GroupSnapshotNamespace```

VictorDenisov replied:
Can it be done is cls_rbd? When we are handling snapshot_add we have access to the image's object. If we want to check whether the image belongs to the group then we need to go to potentially another object on another osd. Is it possible from cls_rbd layer? My understanding is that this only can be done from librbd layer.

dillaman commented on src/librbd/Group.cc:
```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+      ImageCtx *ictx = ictxs[i];
+      ldout(cct, 1) << "Get snap id with name " << ind_snap_name << dendl;
+      ictx->snap_lock.get_read();
+      snap_t snap_id = ictx->get_snap_id(ind_snap_name);
```
If this double-lookup is required, I'd rather have the SnapshotCreateRequest state machine take an optional snap_id pointer to save the assigned snap id into before completing to avoid this lookup.

dillaman commented on src/librbd/Group.cc:
```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+  return ret_code;
+}
+
+int group_snap_list_cls(librados::IoCtx& group_ioctx, const char *group_name,
```
C++ supports overloading -- no need for a unique name

dillaman commented on src/librbd/Group.cc:
```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+  }
+  string group_header_oid = util::group_header_name(group_id);
+
+  r = group_snap_list_cls(group_ioctx, group_name, snaps);
```
Already looked up the id and now ```group_snap_list_cls``` is going to look it up again

dillaman commented on src/librbd/Group.cc:
```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+    return r;
+  }
+  cls::rbd::GroupSnapshot *gs = nullptr;
+  int n = snaps.size();
```
Just need a ```for (auto &gs : groups_names)``` loop and break out when you find a match

VictorDenisov replied:
Sorry, not sure I completely understand what you mean. Is this what you mean:
```
cls::rbd::GroupSnapshot *gs = nullptr;
int n = snaps.size();
for (auto &snap : snaps) {
  if (snap.name == string(snap_name)) {
    gs = &snap;
    break;
  }
}
```

dillaman replied:
Yes

dillaman commented on src/librbd/Group.cc:
```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+      ictxs[i] = nullptr;
+      ret_code = r;
+    }
+    if (ictxs[i]->id != gs->snaps[i].image_id) {
```
Can't rename image ids

dillaman replied:
You should be opening the image via its id and not its name -- no need to lookup the image name.

VictorDenisov replied:
Hm. I thought I couldn't do it with id only. I reread the doc and now I can see it's possible. Thanks.

dillaman commented on src/librbd/Group.cc:
```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+  for (int i = 0; i < n; ++i) {
+    ImageCtx *ictx = ictxs[i];
+    ldout(cct, 20) << "Removing individual snapshot with name: " << image_snap_name << dendl;
+    int r = ictx->state->refresh_if_required();
```
Just opened the image -- no need to refresh (plus it will be handled automatically within ```Operations``` if required)

dillaman commented:
@VictorDenisov Need to update the CLI test to handle the new help text and need unit and workunit tests to exercise the new cls/librbd APIs/rbd CLI.

VictorDenisov commented:
work in progress. I'll update when it's done.

VictorDenisov commented:
@dillaman do I understand correctly that python bindings are in src/pybind/rbd/rbd.pyx?

dillaman commented:
@VictorDenisov Yes, that's correct

dachary commented:
jenkins test this please (eio now ignored in master)

dachary commented:
<pre>
Error compiling Cython file:
------------------------------------------------------------
...
        char *description
        time_t last_update
        bint up

    ctypedef struct rbd_group_image_spec_t:
        char *name;
                 ^
------------------------------------------------------------

rbd.pyx:134:18: Syntax error in C variable declaration
Compiling rbd.pyx because it changed.
Cythonizing rbd.pyx
Traceback (most recent call last):
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/rbd/setup.py", line 188, in <module>
    os.path.join(os.path.dirname(__file__), "..", "rados")
  File "/usr/lib/python2.7/dist-packages/Cython/Build/Dependencies.py", line 798, in cythonize
    cythonize_one(*args[1:])
  File "/usr/lib/python2.7/dist-packages/Cython/Build/Dependencies.py", line 915, in cythonize_one
    raise CompileError(None, pyx_file)
Cython.Compiler.Errors.CompileError: rbd.pyx
</pre>

VictorDenisov commented:
@dillaman I have a question about python bindings. They convert ENOENT and EEXIST error codes into ImageExists and ImageNotFound exceptions. I'm thinking about renaming them to ObjectExists and ObjectNotFound, since the same errors can occur with groups, but this change will break backward compatibility. What is your opinion on it?

VictorDenisov commented:
@dillaman So far I decided to rename the Exceptions to ObjectNotFound. If you don't like it I'll rename back. This PR is ready for review. This PR doesn't have snapshot clone operations for creating consistency groups from it and renaming groups and snapshots. I would prefer to have the in a separate pull request unless you insist.

dillaman commented:
@VictorDenisov I'd prefer not to break backwards compatibility by changing the class names. You could always just directly throw your own exception in those two cases in the group_create/remote functions

VictorDenisov commented:
Ok. I'll change it.

VictorDenisov commented:
Done.

VictorDenisov commented:
@dillaman I presume you are busy with Kraken. I just wanted to make sure that you noticed the update.

dillaman commented:
@VictorDenisov Thanks -- once Kraken is released, I plan to get back to the Luminous features.

VictorDenisov commented:
I'm reworking this code. I'll update when it's done.

dillaman commented:
retest this please

dillaman commented:
@VictorDenisov is this ready for review?

VictorDenisov commented:
@dillaman Yes. It's ready.

dillaman commented:
@VictorDenisov I seemed to have missed this before, but please add all new types to ```test/encoding/types.h``` to ensure they are properly registered.

dillaman commented on src/cls/rbd/cls_rbd_types.h:
```diff
@@ -248,13 +253,13 @@ struct GroupSnapshotNamespace {
   int64_t group_pool;
   string group_id;
-  snapid_t snapshot_id;
+  string snapshot_id;
```
Nit: I think we should either drop the "group_" prefix from all the variables (since it's implied w/ the class name) or be consistent and add it here. I am leaning towards the first choice since it's less verbose.

VictorDenisov replied:
After trying to rename it I realized that it looks confusing from user's perspective.
If I have code in librbd/internal.cc like:
```
rados.ioctx_create2(snap_namespace.group_pool, group_ioctx);
```
or 
```
string group_header_oid = util::group_header_name(snap_namespace.group_id);
```
When it only has id instead of group_id then the impression is that it's a namespace id rather then a group id that this namespace references. In this regard I would prefer to add group to snapshot_id.

dillaman commented on src/include/rbd/librbd.h:
```diff
@@ -146,7 +146,7 @@ typedef enum {
 typedef struct {
   char *name;
-  int64_t pool;
+  char *pool;
```
Why the change to a C-string? I think it's safe to let the caller decode the pool name -- plus, the pool id is actually more helpful if trying to create IoCtx's because it's safer from races.

dillaman commented on src/include/rbd/librbd.h:
```diff
@@ -156,10 +156,20 @@ typedef struct {
 typedef struct {
   char *name;
-  int64_t pool;
+  char *pool;
```
Same comment here

dillaman commented on src/include/rbd/librbd.h:
```diff
@@ -758,6 +768,15 @@ CEPH_RBD_API void rbd_group_image_status_cleanup(
+CEPH_RBD_API int rbd_group_snap_list(rados_ioctx_t group_p, const char *group_name,
+				     rbd_group_snap_spec_t *snaps, size_t *snaps_size);
+CEPH_RBD_API void rbd_group_snap_spec_cleanup(rbd_group_snap_spec_t *group_snap_spec);
+CEPH_RBD_API void rbd_group_snap_spec_list_cleanup(rbd_group_snap_spec_t *snaps,
```
Nit: rename to ```rbd_group_snap_list_cleanup```

dillaman commented on src/include/rbd/librbd.h:
```diff
@@ -758,6 +768,15 @@ CEPH_RBD_API void rbd_group_image_status_cleanup(
+					const char *snap_name);
+CEPH_RBD_API int rbd_group_snap_list(rados_ioctx_t group_p, const char *group_name,
+				     rbd_group_snap_spec_t *snaps, size_t *snaps_size);
+CEPH_RBD_API void rbd_group_snap_spec_cleanup(rbd_group_snap_spec_t *group_snap_spec);
```
What would this be used for?

VictorDenisov replied:
You mean it shouldn't be exposed as a public API? This method is used by rbd_group_snap_spec_list_cleanup. Python bindings only use rbd_group_snap_spec_list_cleanup, I just thought I would add this method for consistency. If we can cleanup a list of snapshot specs, we should be able to cleanup an individual snapshot spec. Do you still want me to remove this method?

dillaman replied:
Correct --- I don't see a use-case for cleaning up just one structure vs freeing up the entire list you just retrieved via the _list method.

VictorDenisov replied:
Fair enough. I'll inline it in rbd_group_snap_spec_list_cleanup.

dillaman commented on src/include/rbd/librbd.hpp:
```diff
@@ -15,14 +15,18 @@
 #include <list>
 #include <map>
 #include <vector>
+#include "common/Formatter.h"
```
This is not part of the public API -- neither are the other two new includes, so you cannot add them.

dillaman commented on src/include/rbd/librbd.hpp:
```diff
@@ -35,9 +39,51 @@ namespace librbd {
     uint64_t id;
     uint64_t size;
     std::string name;
+    snap_namespace_t snap_namespace;
```
Cannot change the existing API -- this would break ABI and crash older clients using the newer library. You will just need to add a new helper method to retrieve the namespace of individual snapshots by id.

dillaman replied:
Since you will have to support the C API as well, and you might as well keep the interfaces semi-similar, perhaps add a new snapshot namespace enum, a new ```snap_get_namespace_type``` call that will retrieve the namespace enum, and a new ```snap_get_group_spec``` (or similar) method to retrieve the group information should the namespace type by a group.

dillaman commented on src/include/rbd/librbd.hpp:
```diff
@@ -35,9 +39,51 @@ namespace librbd {
+    }
+  } group_snap_namespace_t;
+
+  typedef boost::variant<user_snap_namespace_t, group_snap_namespace_t, unknown_snap_namespace_t> snap_namespace_t;
```
Cannot use a variant since future additions would break the ABI

dillaman commented on src/include/rbd/librbd.hpp:
```diff
@@ -75,15 +121,22 @@ namespace librbd {
   typedef struct {
     std::string name;
-    int64_t pool;
+    std::string pool;
```
Same comment about keeping the pool id

dillaman commented on src/include/rbd/librbd.hpp:
```diff
@@ -75,15 +121,22 @@ namespace librbd {
   typedef struct {
     std::string name;
-    int64_t pool;
+    std::string pool;
```
Same comment about keeping the pool id

VictorDenisov replied:
I presume it's also applicable to group_snap_namespace_t in this file, isn't it?

dillaman replied:
Yup -- but I skipped over it since I knew you would have to refactor it to eliminate the variant.

dillaman commented on src/librbd/ImageCtx.h:
```diff
@@ -68,6 +68,7 @@ namespace librbd {
                                         // a format librados can understand
     std::map<librados::snap_t, SnapInfo> snap_info;
     std::map<std::string, librados::snap_t> snap_ids;
+    std::map<cls::rbd::SnapshotNamespace, librados::snap_t> snap_namespace_ids;
```
Please merge ```snap_ids ``` w/ ```snap_namespace_ids```. Use a std::pair of the namespace + the snapshot name for the unique key.

VictorDenisov replied:
I tried merging snap_ids and snap_namespace_ids when I just started working on this and it required so much refactoring of the existing code that I decided to do it in a separate PR if at all. Here the issues I've encountered. Probably, you'll be able to repel my fears. If I use pair <SnapshotNamespace, String> as the key for each entry then I can't search by name only? It means that everywhere where we want to determine the id of a snapshot by its name we need to specify its namespace as well. It means changing even command line tools in order to allow the user to specify what namespace he wants to look in. Which seems to me like a too big change.

If we use reversed pair <string, SnapshotNamespace> then we can look for snapshots using only their names, but then we can't look for snapshots using only their Namespace. But this is why we want this index in the first place. Taking all of this into account I decided to just have a separate index for namespaces. What is your thought on this?

dillaman replied:
I am probably missing something, but you would need to do this regardless since you can have duplicate names between namespaces. There really would never be a valid case where you need to search by name only since you would need the namespace disambiguation. For the legacy snapshots, you would just default construct an image namespace and the snapshot name. For group snapshots, you would require the full group namespace to find the correct snapshot.

I would actually recommend breaking this piece of refactoring into its own commit within the PR since it is an atomic unit of work that needs to be done and would facilitate the review.

VictorDenisov replied:
Should we allow users to rename snapshots from GroupSnapshotNamespace?

VictorDenisov replied:
I'll do this refactoring in a separate branch and will show you. If you like it then I'll rebase this PR on top of the refactoring branch.

VictorDenisov replied:
@dillaman I created another PR #12970 for this refactoring. I would like to move through that PR gradually. Could you please take a look at the current version. It has only changes to ImageCtx. Please take a look and tell me if this is what you want to do to ImageCtx in principle. Thanks.

VictorDenisov replied:
The PR is very short yet. It shouldn't take long to review.

dillaman commented on src/tools/rbd/ArgumentTypes.h:
```diff
@@ -45,6 +45,7 @@ static const std::string JOURNAL_SPEC("journal-spec");
 static const std::string PATH_NAME("path-name");
 
 // optional arguments
+static const std::string ALL_NAME("all");
```
Nit: just declare/define this new argument within Snap.cc since that's the only place it's currently used. The fields are are re-used between other actions.

dillaman commented:
@VictorDenisov Can you create a simple script in "qa/workunits/rbd" that tests all the new group commands available from the rbd CLI (i.e. create/delete groups, attach/detach images, create/remove snapshots, etc)?

VictorDenisov commented:
Sure. I'll add the test.

VictorDenisov commented:
@dillaman What kind of types should be added to test/encoding/types.h . I looked at types from cls_rbd_types.h. It looks like all enums are not added. Also struct MirrorImageStatus is not added. Is there something special about MirrorImageStatus or was it just overlooked?

dillaman commented:
@VictorDenisov Basically, any non-POD type w/ an encode/decode should be added to that test. In the case of MirrorImageStatus, it was overlooked.

VictorDenisov commented on src/cls/rbd/cls_rbd_types.h:
```diff
@@ -310,6 +329,52 @@ struct SnapshotNamespaceOnDisk {
+  std::string name;
+  GroupSnapshotState state;
+
+  vector<ImageSnapshotRef> snaps;
```
Images can be removed and added to a group. Snapshots should stay intact. This list is primarily for keeping the list of images. Snapshot id in this reference comes in handy when we need to list image snapshots in this group snapshot.

dillaman replied:
How does OpenStack behave? It seems odd to me to remove an image from a group w/o also deleting the group's associated snapshots within the image. Since we only permit group snapshots to be manipulated via the group, that seems like an oddity (i.e. the image is not part of a group anymore, but you cannot delete the image w/o deleting the linked group snapshots -- which would affect all images still within the group).

VictorDenisov replied:
According to my conversation with Xing, images can be removed and added. Here is the link to my email thread with him: http://lists.openstack.org/pipermail/openstack-dev/2016-November/106958.html

VictorDenisov commented:
Work in progress. I'll update when it's done.

VictorDenisov commented:
These changes are rebased onto master, but they still require some comments to be addressed and some commits to be squashed. I'll update when it's ready.

VictorDenisov commented:
@dillaman How should those qa/workunits tests be invoked?

dillaman commented:
@VictorDenisov You can add new workunit test cases in this directory following the example of existing tasks [1]

[1] https://github.com/ceph/ceph/tree/master/qa/suites/rbd/singleton/all

VictorDenisov commented:
@dillaman  could you please take a look at the change. I addressed the comments. Ceph-test-helpers keeps failing, but it's not clear to me how it's related to my changes.

VictorDenisov commented:
rebase on the latest master fixed the issue.

dillaman commented on src/librbd/api/Group.cc:
```diff
@@ -26,20 +30,280 @@ using librados::Rados;
+  int snap_count = group_snap.snaps.size();
+
+  for (int i = 0; i < snap_count; ++i) {
+    librbd::IoCtx* image_io_ctx = new librbd::IoCtx;
```
Unnecessary to allocate on the heap -- it internally reference counts (plus ```ImageCtx``` just creates its own immediately)

dillaman commented on src/librbd/api/Group.cc:
```diff
@@ -26,20 +30,280 @@ using librados::Rados;
+}
+
+static int group_image_list(librados::IoCtx& group_ioctx, const char *group_name,
+    std::vector<cls::rbd::GroupImageStatus> *image_ids)
```
Nit: indentation

dillaman commented on src/tools/rbd/action/Snap.cc:
```diff
@@ -29,13 +43,23 @@ int do_list_snaps(librbd::Image& image, Formatter *f)
+    snaps.erase(remove_if(snaps.begin(),
+			  snaps.end(),
+			  boost::bind(is_not_user_snap_namespace, &image, _1)),
+		snaps.end());
```
Nit: indentation

VictorDenisov replied:
This indentation is meant to show that snaps.end() is erase's argument as well as remove_id. Do you want me to join it all in one line and wrap based on line length?

dillaman replied:
OK -- disregard then.

dillaman commented on src/tools/rbd/action/Snap.cc:
```diff
@@ -48,17 +72,40 @@ int do_list_snaps(librbd::Image& image, Formatter *f)
+      if (all_snaps) {
+	f->open_object_section("namespace");
+	if (get_group_res == 0) {
+	  f->dump_int("pool", group_snap.group_pool);
```
Nit: translate pool id to pool name

dillaman commented on src/tools/rbd/action/Snap.cc:
```diff
@@ -48,17 +72,40 @@ int do_list_snaps(librbd::Image& image, Formatter *f)
+      std::string namespace_string;
+      if (all_snaps && (get_group_res == 0)) {
+	ostringstream oss;
+	oss << "group snapshot member - " << group_snap.group_pool << "/" <<
```
Nit: translate pool id to pool name

dillaman replied:
Nit: can probably shorten "group snapshot member - " to just "group snapshot"

dillaman commented on src/librbd/internal.h:
```diff
@@ -195,6 +195,39 @@ namespace librbd {
   int metadata_list(ImageCtx *ictx, const string &last, uint64_t max, map<string, bufferlist> *pairs);
   int metadata_get(ImageCtx *ictx, const std::string &key, std::string *value);
 
+  int mirror_mode_get(IoCtx& io_ctx, rbd_mirror_mode_t *mirror_mode);
```
Rebase issue here -- pulled in lots of deleted lines

dillaman commented on src/librbd/librbd.cc:
```diff
@@ -3796,3 +3853,92 @@ extern "C" void rbd_group_image_status_list_cleanup(
+    group_snap_spec_cpp_to_c(cpp_snaps[i], &snaps[i]);
+  }
+
+  r = cpp_snaps.size();
```
Nit: return zero upon success and always update ```snap_size``` w/ the correct count

VictorDenisov replied:
I can assign this to snaps_size, but it looks like all list functions return list size in case of success, not zero.

dillaman replied:
The C API is half-and-half -- I'd prefer to return zero upon success and avoid signed vs unsigned int issues

dillaman commented on src/cls/rbd/cls_rbd_types.h:
```diff
@@ -328,6 +336,68 @@ struct SnapshotNamespaceOnDisk {
+SnapshotNamespaceType get_namespace_type(const SnapshotNamespace& snapshot_namespace);
+
+enum GroupSnapshotState {
+  GROUP_SNAPSHOT_STATE_PENDING = 0,
```
Nit: can we rename this to ```GROUP_SNAPSHOT_STATE_INCOMPLETE```? If there is ever an issue, a PENDING group snapshot will never migrate to COMPLETE.

dillaman commented on src/pybind/rbd/rbd.pyx:
```diff
@@ -2553,3 +2841,98 @@ cdef class SnapIterator(object):
             rbd_snap_list_end(self.snaps)
             free(self.snaps)
+
+cdef class ImageIterator(object):
```
Nit: rename to ```GroupImageIterator```

dillaman commented on qa/workunits/rbd/rbd_consistency_groups.sh:
```diff
@@ -0,0 +1,45 @@
+#!/bin/sh -ex
```
Nit: rename this to "rbd_groups.sh"

dillaman commented on qa/suites/rbd/cli/workloads/rbd_cli_consistency_groups.yaml:
```diff
@@ -0,0 +1,5 @@
+tasks:
```
Nit: rename this to "rbd_cli_groups.yaml"

dillaman commented on qa/workunits/rbd/rbd_consistency_groups_helpers.sh:
```diff
@@ -0,0 +1,136 @@
+#!/bin/sh
```
Nit: combine this w/ rbd_groups.sh -- a separate file isn't much use when you don't have multiple tests referencing common functions

dillaman commented on src/librbd/internal.cc:
```diff
@@ -2056,6 +2056,73 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
+    const SnapInfo *snap_info;
+    {
+      RWLock::RLocker l(ictx->snap_lock);
+      snap_info = ictx->get_snap_info(snap_id);
```
Check for nullptr -- cannot reference ```snap_info``` outside lock scope

dillaman commented on src/librbd/internal.cc:
```diff
@@ -2056,6 +2056,73 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
+    }
+  };
+
+  int snap_get_group(ImageCtx *ictx,
```
Nit: I would move this to api/Group

dillaman commented on src/librbd/internal.cc:
```diff
@@ -2056,6 +2056,73 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
+    const SnapInfo *snap_info;
+    {
+      RWLock::RLocker l(ictx->snap_lock);
+      snap_info = ictx->get_snap_info(snap_id);
```
Same comment here

dillaman commented on src/librbd/internal.cc:
```diff
@@ -2056,6 +2056,73 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
+    return 0;
+  }
+
+  class GetGroupVisitor : public boost::static_visitor<void> {
```
Make this a "int" visitor so that you can return your result code

dillaman replied:
Nit: move to api/Group within an anonymous namespace

dillaman commented on src/librbd/internal.cc:
```diff
@@ -2056,6 +2056,73 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
+  public:
+    librados::IoCtx *image_ioctx;
+    group_snap_t group_snap;
+    bool found = false;
```
No need for these two variables once you can directly return your result code

dillaman commented on src/librbd/internal.cc:
```diff
@@ -2056,6 +2056,73 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
+      found = true;
+      librados::Rados rados(*image_ioctx);
+      IoCtx group_ioctx;
+      rados.ioctx_create2(snap_namespace.group_pool, group_ioctx);
```
Check for success

dillaman commented on src/librbd/internal.cc:
```diff
@@ -2056,6 +2056,73 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
+      std::string group_name;
+      r = cls_client::dir_get_name(&group_ioctx, RBD_GROUP_DIRECTORY,
+				   snap_namespace.group_id, &group_name);
 
```
Return result upon failure

dillaman commented on src/librbd/internal.cc:
```diff
@@ -2056,6 +2056,73 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
+					   group_header_oid,
+					   snap_namespace.group_snapshot_id,
+					   &group_snapshot);
+
```
Return result upon failure

dillaman commented on src/librbd/internal.cc:
```diff
@@ -2056,6 +2056,73 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
+  class GetGroupVisitor : public boost::static_visitor<void> {
+  public:
+    librados::IoCtx *image_ioctx;
+    group_snap_t group_snap;
```
Just pass the pointer so you can directly update

dillaman commented on src/librbd/api/Group.cc:
```diff
@@ -318,80 +628,273 @@ int Group<I>::image_list(librados::IoCtx& group_ioctx,
-    librados::Rados rados(ictx->md_ctx);
-    IoCtx ioctx;
-    rados.ioctx_create2(ictx->group_spec.pool_id, ioctx);
+  r = group_snap_list(group_ioctx, group_name, &snaps);
```
Good use for a ```group_snap_get_by_name``` function. That would allow the removal of cls_rbd::check_duplicate_snap_name because it can just use the a helper "get_by_name" function

dillaman commented on src/librbd/api/Group.cc:
```diff
@@ -26,20 +30,280 @@ using librados::Rados;
 
-template <typename I>
-int Group<I>::create(librados::IoCtx& io_ctx, const char *group_name)
+static string generate_uuid(librados::IoCtx& io_ctx)
```
Put all static functions in an anonymous namespace near the top of the file

VictorDenisov replied:
I presume static keyword should be removed.

dillaman replied:
Yes, the ```static``` keyword is not required when placed in an anonymous namespace

dillaman commented:
@VictorDenisov looking close -- can you remove any newly added "consistency" verbiage references since we are otherwise just referring to this functionality as groups.

VictorDenisov commented:
Yep. Will do. Most likely will be ready after the weekend.

VictorDenisov commented:
I was too optimistic. I'll update when I'm done.

VictorDenisov commented:
@dillaman ready for another review.

dillaman commented:
@VictorDenisov Looks like lots of related test failures:
http://pulpito.ceph.com/jdillaman-2017-05-15_11:02:46-rbd-wip-jd-testing-distro-basic-smithi/

dillaman commented:
@VictorDenisov ... one more thing: it looks like your recent commits are missing the "librbd" prefix. In fact, they can probably just be squashed down to the commit for the comment they are addressing.

dillaman commented on src/tools/rbd/action/Snap.cc:
```diff
@@ -220,6 +281,13 @@ int execute_list(const po::variables_map &vm) {
     return r;
   }
+  bool all_snaps = vm[ALL_NAME].as<bool>();
+  r = utils::get_pool_image_snapshot_names(
```
Looks like this is a rebase error

VictorDenisov commented:
@dillaman How can I trigger this test run(http://pulpito.ceph.com/jdillaman-2017-05-15_11:02:46-rbd-wip-jd-testing-distro-basic-smithi/) again and see the results?

dillaman commented:
@VictorDenisov The majority of the tests you can run locally if you just start a cluster via "vstart" and manually run the workunit test.

VictorDenisov commented:
@dillaman I fixed python tests, but I don't understand why rbd_groups.sh fails. It works for me locally. Is it possible that you have some other changes in your wip branch that interfere with my code?

dillaman commented:
@VictorDenisov It might have been a rebase error that you have since corrected. Let me attempt to run it locally.

dillaman commented:
@VictorDenisov The rbd_group test fails for me locally if I don't override the ```rbd default features``` and also if I use the data pool feature.

VictorDenisov commented:
I see. I'll look into it. Thanks,

VictorDenisov commented:
It looks like the problem is that I open images using their ids instead of names. Opening images by ids has a completely different workflow. I'm still looking into it, just wanted to share my findings.

VictorDenisov commented:
@dillaman I fixed the problem with datapool. It should be good for another run of teuthology.

dillaman commented:
@VictorDenisov ack -- thanks

Songweibin commented:
@dillaman  @VictorDenisov  Hi, is this PR will merge into the version of Luminous? Thanks.

dillaman commented:
@Songweibin I think mimic is the most likely release

VictorDenisov commented:
@dillaman I rebased the changes. Let me know if anything else is required to merge this PR.

dillaman commented:
@VictorDenisov thanks -- I'll give it a quick once-over tomorrow and re-run it through testing.

Songweibin commented:
Hi, @VictorDenisov 
I tested this in my local and got an rbd error as follow:
```
root@ceph:/ceph-dev/build# ./bin/rbd snap list i1 -a
SNAPID NAME                                            SIZE TIMESTAMP                NAMESPACE                  
     4 0000000000000003_103974b0dc51_103f3d1b58ba 102400 kB Tue Nov 28 18:25:02 2017 group snapshot - rbd/g1@s1 
root@ceph:/ceph-dev/build# ./bin/rbd du i1          
2017-11-28 18:37:35.588 7f9e977fe700 -1 librbd::image::OpenRequest: failed to set image snapshot: (2) No such file or directory
rbd: error opening snapshot i1@0000000000000003_103974b0dc51_103f3d1b58ba: (2) No such file or directory
NAME PROVISIONED USED
```

Songweibin commented on src/tools/rbd/action/Group.cc:
```diff
@@ -287,20 +288,19 @@ int execute_list_images(const po::variables_map &vm) {
       f->dump_int("state", state);
     } else
-      std::cout << pool_id << "." << image_name << " " << state_string << std::endl;
+      std::cout << image.pool << "/" << image_name << " " << state_string << std::endl;
```
@VictorDenisov  Why not present `pool_name/image_name` which i think is more readable than `pool_id/image_name`, thanks.

dillaman commented:
@VictorDenisov Similar issue: I run "rbd group rm <name>" on a group w/ snapshots and I see "librbd::Operations: No such snapshot found." errors (and the snapshots are not properly deleted from the image).

dillaman commented:
Unrelated to these changes, but ```execute_add``` and ```execute_remove_image``` in ```src/tools/rbd/action/Group.cc``` both leak memory due to the second use of ```util::init``` when the second call should be ```util::init_io_ctx``` instead.

dillaman commented on src/cls/rbd/cls_rbd_types.h:
```diff
@@ -328,6 +336,67 @@ struct SnapshotNamespaceOnDisk {
+struct GroupSnapshot {
+  std::string id;
+  std::string name;
+  GroupSnapshotState state;
```
Nit: default initialize to a valid value

dillaman commented on src/tools/rbd/action/Group.cc:
```diff
@@ -396,6 +588,15 @@ Shell::Action action_remove_image(
+  {"group", "snap", "create"}, {}, "Make a snapshot of a group.",
+  "", &get_group_snap_create_arguments, &execute_group_snap_create);
+Shell::Action action_group_snap_remove(
+  {"group", "snap", "remove"}, {}, "Remove a snapshot from a group.",
```
Nit: add "rm" alias for "remove" for CLI consistency

VictorDenisov commented on src/cls/rbd/cls_rbd.h:
```diff
@@ -164,4 +164,15 @@ struct cls_rbd_snap {
 };
 WRITE_CLASS_ENCODER(cls_rbd_snap)
 
+namespace group {
```
@dillaman I need this function to be available in unit tests: ceph/src/test/cls_rbd/test_cls_rbd.cc:2308:21. Since the test was not about building snap_key I didn't want to duplicate this code. What is your opinion?

dillaman replied:
OK -- leave it then.

VictorDenisov commented on src/include/rbd/librbd.h:
```diff
@@ -894,6 +910,18 @@ CEPH_RBD_API void rbd_group_image_status_cleanup(
+				     rbd_group_snap_spec_t *snaps, size_t *snaps_size);
+CEPH_RBD_API void rbd_group_snap_list_cleanup(rbd_group_snap_spec_t *snaps,
+						   size_t len);
+
```
Can I add these features as part of this PR: https://github.com/ceph/ceph/pull/12637

dillaman replied:
I'd prefer it to be part of this one to (1) avoid getting lost and (2) since this one added the functions.

liewegas commented:
@dillaman what's the status of this one?

VictorDenisov commented:
I'm addressing the comments and fixing one bug.

dillaman commented:
Manually resolved remaining issues and merged to master branch
